Thrift communication errors -- heavy exception flood

There are certain cases when Thrift starts to generate heavy `COMMUNICATION_ERROR`s.
# Cases
-  If an _entirely_ new project is added &mdash; this means **no** CodeChecker nature.
-  If an project without CodeChecker nature gets built
- If a project **with** CodeChecker nature but **without** a present run is accessed by various means (file change, project build, etc.)

There might be more causes, these are just the list of those discovered so far!

---

It seems like that all these issues boil down to the fact that the codes calling Thrift **assumes** that
- A nature is added
- A server is running
- The server has been configured in a working fashion
- A CodeChecker run has been already created

These are all assumptions that are **NOT** invariant to every project, so these assumptions should be eliminated from the code. In the event these conditions do not hold true, access to the server should not be sent.

Maybe it's an issue with CodeChecker itself, that it does not respond in an orderly fashion when I try to get results for something nonexistent? I'm summoning @gyorb and @Xazax-hun to this discussion.

---
# Example

The exceptions look something like this:

> 2016-08-16 13:16:49 Thread-14  ERROR ThriftActionImpl:35 - Error in cc.ecl.action.thrift transport creation
> org.apache.thrift.transport.TTransportException: java.net.ConnectException: Connection refused
>         at org.apache.thrift.transport.THttpClient.flush(THttpClient.java:356)
>         at org.apache.thrift.TServiceClient.sendBase(TServiceClient.java:65)
>         at cc.codechecker.service.thrift.gen.CodeCheckerDBAccess$Client.send_getRunData(CodeCheckerDBAccess.java:163)
>         at cc.codechecker.service.thrift.gen.CodeCheckerDBAccess$Client.getRunData(CodeCheckerDBAccess.java:156)
>         at cc.codechecker.api.thrift.run.list.ListRunsActionThriftImpl.runThrift(ListRunsActionThriftImpl.java:38)
>         at cc.codechecker.api.thrift.run.list.ListRunsActionThriftImpl.runThrift(ListRunsActionThriftImpl.java:22)
>         at cc.ecl.action.thrift.ThriftActionImpl.doRealRun(ThriftActionImpl.java:33)
>         at cc.ecl.action.thrift.ThriftActionImpl.doRealRun(ThriftActionImpl.java:17)
>         at cc.ecl.action.AbstractActionImpl.doRun(AbstractActionImpl.java:44)
>         at cc.ecl.action.Action.runWithImpl(Action.java:64)
>         at cc.ecl.action.SimpleActionRunner.requireResult(SimpleActionRunner.java:152)
>         at cc.codechecker.api.thrift.result.count.ResultCountActionThriftImpl.runThrift(ResultCountActionThriftImpl.java:54)
>         at cc.codechecker.api.thrift.result.count.ResultCountActionThriftImpl.runThrift(ResultCountActionThriftImpl.java:23)
>         at cc.ecl.action.thrift.ThriftActionImpl.doRealRun(ThriftActionImpl.java:33)
>         at cc.ecl.action.thrift.ThriftActionImpl.doRealRun(ThriftActionImpl.java:17)
>         at cc.ecl.action.AbstractActionImpl.doRun(AbstractActionImpl.java:44)
>         at cc.ecl.action.Action.runWithImpl(Action.java:64)
>         at cc.ecl.action.ActionQueue$QueueItem.runAction(ActionQueue.java:89)
>         at cc.ecl.action.SimpleActionRunner.runAction(SimpleActionRunner.java:53)
>         at cc.ecl.action.SimpleActionRunner.run(SimpleActionRunner.java:98)
>         at java.lang.Thread.run(Thread.java:745)
> Caused by: java.net.ConnectException: Connection refused
>         at java.net.PlainSocketImpl.socketConnect(Native Method)
>         at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
>         at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
>         at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
>         at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
>         at java.net.Socket.connect(Socket.java:589)
>         at java.net.Socket.connect(Socket.java:538)
>         at sun.net.NetworkClient.doConnect(NetworkClient.java:180)
>         at sun.net.www.http.HttpClient.openServer(HttpClient.java:432)
>         at sun.net.www.http.HttpClient.openServer(HttpClient.java:527)
>         at sun.net.www.http.HttpClient.<init>(HttpClient.java:211)
>         at sun.net.www.http.HttpClient.New(HttpClient.java:308)
>         at sun.net.www.http.HttpClient.New(HttpClient.java:326)
>         at sun.net.www.protocol.http.HttpURLConnection.getNewHttpClient(HttpURLConnection.java:1169)
>         at sun.net.www.protocol.http.HttpURLConnection.plainConnect0(HttpURLConnection.java:1105)
>         at sun.net.www.protocol.http.HttpURLConnection.plainConnect(HttpURLConnection.java:999)
>         at sun.net.www.protocol.http.HttpURLConnection.connect(HttpURLConnection.java:933)
>         at org.apache.thrift.transport.THttpClient.flush(THttpClient.java:344)
>         ... 20 more

--- a/eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/config/CcConfiguration.java
+++ b/eclipse-plugin/eclipse/cc.codechecker.eclipse.plugin/src/cc/codechecker/plugin/config/CcConfiguration.java
@@ -113,7 +113,7 @@ public class CcConfiguration {
         try {
             return CodeCheckerContext.getInstance().getServerObject(project).getServerUrl();
         } catch (Exception e) {
-            return "http://localhost:11444/";
+            return "";
         }
     }
 
@@ -210,7 +210,18 @@ public class CcConfiguration {
     }
 
     public boolean isConfigured() {
-        return !getServerUrl().equals("");
+        try {
+            if(project.hasNature(CodeCheckerNature.NATURE_ID)) {
+                CodeCheckEnvironmentChecker ccec = CodeCheckerContext.getInstance().getServerObject(project).getCodecheckerEnvironment();
+                if(ccec != null) {
+                    return ccec.isJavaRunner(CodeCheckerContext.getInstance().getServerObject(project).serverPort);
+                }
+                return false;
+            }
+            return false;
+        } catch (CoreException e) {
+            return false;
+        }
     }
 
     public void updateServer(IProject project, CodecheckerServerThread server) {

--- a/eclipse-plugin/service/cc.codechecker.service.api/src/main/java/cc/codechecker/api/runtime/CodeCheckEnvironmentChecker.java
+++ b/eclipse-plugin/service/cc.codechecker.service.api/src/main/java/cc/codechecker/api/runtime/CodeCheckEnvironmentChecker.java
@@ -100,6 +100,17 @@ public class CodeCheckEnvironmentChecker {
                         other.workspaceName);
     }
     
+    public boolean isJavaRunner(int serverPort) {
+        ShellExecutorHelper she = new ShellExecutorHelper(environmentBefore);
+
+        String cmd = codeCheckerCommand + " cmd runs -p " + serverPort + " -o json";
+        Optional<String> ccOutput = she.waitReturnOutput(cmd);
+        if (ccOutput.isPresent()) {
+            return ccOutput.get().contains("javarunner");
+        }
+        return false;
+    }
+
     public void setCheckerCommand(String checkerCommand) {
         this.checkerCommand = checkerCommand;
     }

